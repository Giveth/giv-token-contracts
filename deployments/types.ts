export interface IUnipoolContractConfig {
    uniTokenAddress: string; // The LP token address
    // Duration of each reward period which start after calling notify reward on
    // unipool
    lmDuration: number;
    // The amount of tokens will be paid as reward in reward period (i.e. lmDuration)
    rewardAmount: string;
}
export interface IRegenConfig {
    // This field can be used to reuse already deployed tokenDistro and not deploy
    // a new one
    alreadyDeployedTokenDistroAddress?: string;
    newTokenDistroParams?: {
        // Token distro start time (i.e. initial percentage releases at this moment
        // and the stream begins after cliff time
        startTime: number;
        // Time period between token distro start (initial amount release)
        // and stream start
        cliffPeriod: number;
        // Total duration of token distro (cliff period + streaming period)
        // Note: 100% of allocated tokens will be release at startTime+duration
        duration: number;
        // The percentage of tokens release immediately at startTime
        initialPercentage: number;
        // The address of token will be distributed by token distro
        tokenAddress: string;
        // Total number of tokens can be distributed
        totalTokens: string;
        // Whether admins can cancel stream allocated to an account and
        // transfer it to another one
        cancelable: boolean;
    };
    unipools: {
        [key: string]: IUnipoolContractConfig;
    };
}
