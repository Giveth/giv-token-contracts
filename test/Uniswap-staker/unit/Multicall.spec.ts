import { LoadFixtureFunction } from "../types";
import { describe, beforeEach, before, it } from "mocha";
import {
    uniswapFixture,
    mintPosition,
    UniswapFixtureType,
} from "../shared/fixtures";
import {
    getMaxTick,
    getMinTick,
    FeeAmount,
    TICK_SPACINGS,
    blockTimestamp,
    BN,
    BNe18,
    snapshotGasCost,
    ActorFixture,
    makeTimestamps,
    maxGas,
    defaultTicksArray,
    expect,
} from "../shared";
import { createFixtureLoader, provider } from "../shared/provider";
import {
    HelperCommands,
    ERC20Helper,
    incentiveResultToStakeAdapter,
} from "../helpers";
import { createTimeMachine } from "../shared/time";
import { HelperTypes } from "../helpers/types";

let loadFixture: LoadFixtureFunction;

describe("unit/Multicall", () => {
    const actors = new ActorFixture(provider.getWallets(), provider);
    const incentiveCreator = actors.incentiveCreator();
    const lpUser0 = actors.lpUser0();
    const amountDesired = BNe18(10);
    const totalReward = BNe18(100);
    const erc20Helper = new ERC20Helper();
    const Time = createTimeMachine(provider);
    let helpers: HelperCommands;
    let context: UniswapFixtureType;
    const multicaller = actors.traderUser2();

    before("loader", async () => {
        loadFixture = createFixtureLoader(provider.getWallets(), provider);
    });

    beforeEach("create fixture loader", async () => {
        context = await loadFixture(uniswapFixture);
        helpers = HelperCommands.fromTestContext(context, actors, provider);
    });

    it("is implemented", async () => {
        const currentTime = await blockTimestamp();

        await erc20Helper.ensureBalancesAndApprovals(
            multicaller,
            [context.token0, context.token1],
            amountDesired,
            context.nft.address,
        );
        await mintPosition(context.nft.connect(multicaller), {
            token0: context.token0.address,
            token1: context.token1.address,
            fee: FeeAmount.MEDIUM,
            tickLower: getMinTick(TICK_SPACINGS[FeeAmount.MEDIUM]),
            tickUpper: getMaxTick(TICK_SPACINGS[FeeAmount.MEDIUM]),
            recipient: multicaller.address,
            amount0Desired: amountDesired,
            amount1Desired: amountDesired,
            amount0Min: 0,
            amount1Min: 0,
            deadline: currentTime + 10_000,
        });

        // It's not required for UniswapV3RewardToken
        // await erc20Helper.ensureBalancesAndApprovals(
        //     multicaller,
        //     context.rewardToken,
        //     totalReward,
        //     context.staker.address,
        // );

        await context.rewardToken
            .connect(incentiveCreator)
            .transferOwnership(multicaller.address);

        const createIncentiveTx = context.staker.interface.encodeFunctionData(
            "createIncentive",
            [
                {
                    pool: context.pool01,
                    rewardToken: context.rewardToken.address,
                    refundee: incentiveCreator.address,
                    ...makeTimestamps(currentTime + 100),
                },
                totalReward,
            ],
        );
        await context.staker
            .connect(multicaller)
            .multicall([createIncentiveTx], maxGas);

        // expect((await context.staker.deposits(tokenId)).owner).to.eq(
        //   multicaller.address
        // )
    });
});
