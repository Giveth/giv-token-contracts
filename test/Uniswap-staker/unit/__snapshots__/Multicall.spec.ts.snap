// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit/Multicall can be used to exit a position from multiple incentives 1`] = `196446`;

exports[`unit/Multicall can be used to stake an already deposited token for multiple incentives 1`] = `241964`;
